crates/sui-graphql-client/src/faucet.rs

```diff
@@ -0,0 +1,235 @@
+    /// of tokens transferred, and the transaction digest.
+    pub async fn request_and_wait(
+        &self,
+        address: Address,
+    ) -> Result<Option<FaucetReceipt>, anyhow::Error> {
+        let request_id = self.request(address).await?;
+        if let Some(request_id) = request_id {
+            let status_url = Url::parse(
+                self.faucet_url
+                    .to_string()
+                    .replace("gas", "status")
+                    .as_str(),
+            )
+            .expect("Invalid faucet status URL");
+
+            let poll_response = tokio::time::timeout(FAUCET_REQUEST_TIMEOUT, async {
+                let mut interval = tokio::time::interval(FAUCET_POLL_INTERVAL);
+                loop {
+                    interval.tick().await;
+                    eprint!("Polling faucet status...");
+                    let req = self
+                        .request_status(request_id.clone(), status_url.clone())
+                        .await;
+
+                    if let Ok(Some(poll_response)) = req {
+                        eprintln!("response: {:?}", poll_response);
+                        match poll_response.status {
+                            BatchSendStatusType::Succeeded => {
+                                break Ok(poll_response);
+                            }
+                            BatchSendStatusType::Discarded => {
+                                break Ok(BatchSendStatus {
+                                    status: BatchSendStatusType::Discarded,
+                                    transferred_gas_objects: None,
+                                });
+                            }
+                            BatchSendStatusType::Inprogress => {
+                                continue;
+                            }
+                        }
+                    } else if req.is_err() {
+                        break Err(anyhow::anyhow!("Faucet request failed. {:?}", req.err()));
+                    }
+                }
+            })
+            .await
+            .map_err(|_| anyhow::anyhow!("Faucet request timed out"))??;
+            Ok(poll_response.transferred_gas_objects)
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Request gas from a custom faucet service.
+    pub async fn request_url(
+        &self,
+        address: Address,
+        url: Url,
+    ) -> Result<Option<String>, anyhow::Error> {
+        self.request_impl(address, &url).await
+    }
+
+    /// Check the faucet request status.
+    ///
+    /// Possible statuses are defined in: [`BatchSendStatusType`]
+    pub async fn request_status(
+        &self,
+        id: String,
+        url: Url,
+    ) -> Result<Option<BatchSendStatus>, anyhow::Error> {
+        let url = format!("{}/{}", url, id);
+        let response = self.inner.get(&url).send().await?;
+        let json = response.json::<BatchStatusFaucetResponse>().await?;
+        Ok(json.status)
+    }
+}
```

Comment:
same here, use the internal stored url vs passing one in and take id by &str

Reply:
this one is trickier depending on the faucet service (if they're not using our own service). Because you could have different paths for status and requesting a URL, so we'd either need to store both, or allow for the status url to be passed.

Reply:
we need to standardize on *something* so our service, and how the routes are defined, is the purpose of this client is it not?

Reply:
Not sure I follow, you mean we should set the standard (e.g., now we have `v1/gas` and `v1/status`) and custom faucet services should follow it?

Reply:
yes exactly. if it doesn't follow that standard then this faucet implementation won't work either way

Reply:
Fixed